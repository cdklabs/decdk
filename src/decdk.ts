import * as cdk from 'aws-cdk-lib';
import chalk from 'chalk';
import yargs from 'yargs';
import { DeclarativeStack, loadTypeSystem, readTemplate, stackNameFromFileName } from '.';

async function main() {
  const argv = await yargs
    .usage('$0 <filename>', 'Synthesize a CDK stack from a declarative JSON schema')
    .positional('filename', { type: 'string', required: true })
    .argv;


  const templateFile = argv.filename;
  if (!templateFile) {
    throw new Error('filename is missing');
  }

  const template = await readTemplate(templateFile);
  const stackName = stackNameFromFileName(templateFile);
  const typeSystem = await loadTypeSystem();

  const app = new cdk.App();
  new DeclarativeStack(app, stackName, { template, typeSystem });
  app.synth();
}

main().catch(e => {
  // eslint-disable-next-line no-console
  console.error(chalk.red(e));
  process.exit(1);
});
