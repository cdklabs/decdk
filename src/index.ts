export * from './declarative-stack';
export * from './util';
