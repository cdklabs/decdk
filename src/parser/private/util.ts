export function mkDict<A>(
  xs: ReadonlyArray<readonly [string, A]>
): Record<string, A> {
  const ret: Record<string, A> = {};
  for (const [k, v] of xs) {
    ret[k] = v;
  }
  return ret;
}
