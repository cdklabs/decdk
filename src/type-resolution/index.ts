export * from './resource-like';
export { resolveExpressionType } from './resolve';
export * from './template';
