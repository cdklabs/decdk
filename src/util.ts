import * as path from 'path';
import * as jsiiReflect from 'jsii-reflect';
import { Template } from './parser/template';

/**
 * Reads a YAML/JSON template file.
 */
export async function readTemplate(templateFile: string): Promise<Template> {
  return Template.fromFile(templateFile);
}

export async function loadTypeSystem(validate = true) {
  const typeSystem = new jsiiReflect.TypeSystem();
  await typeSystem.loadNpmDependencies(path.resolve(__dirname, '..'), {
    validate,
  });
  return typeSystem;
}

export function stackNameFromFileName(fileName: string) {
  return path.parse(fileName).name.replace('.', '-');
}
