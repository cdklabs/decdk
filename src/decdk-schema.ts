import yargs from 'yargs';
import { renderFullSchema } from './schema/cdk-schema';
import { loadTypeSystem } from './util';

/* eslint-disable no-console */

async function main() {
  const argv = await yargs
    .usage('$0', 'Generate a JSON Schema to validate deCDK templates')
    .option('warnings', {
      default: true,
      type: 'boolean',
    }).argv;

  const typeSystem = await loadTypeSystem();
  const schema = await renderFullSchema(typeSystem, {
    colors: true,
    warnings: argv.warnings,
    suppressWarnings: [
      // Only an object with methods can satisfy this interface and no built-in class is provided since the feature is intended as an escape hatch
      'aws-cdk-lib.aws_lambda_nodejs.BundlingOptions.commandHooks',
      // Only an object with methods can satisfy this and no built-in class is provided since the feature is intended for user land functionality
      'aws-cdk-lib.BundlingOptions.local',
    ],
  });
  console.log(JSON.stringify(schema, undefined, 2));
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
