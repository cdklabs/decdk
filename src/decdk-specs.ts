import { generateDeCDKSpecs } from './specs/specs-generation';
import { loadTypeSystem } from './util';

/* eslint-disable no-console */

async function main() {
  const typeSystem = await loadTypeSystem();
  const specs = await generateDeCDKSpecs(typeSystem);
  console.log(JSON.stringify(specs, undefined, 2));
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
