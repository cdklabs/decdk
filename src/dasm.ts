import * as cdk from 'aws-cdk-lib';
import chalk from 'chalk';
import yargs from 'yargs';
import { AnnotationsContext, DeclarativeStackError } from './error-handling';
import { CodeEvaluationContext, CodeEvaluator } from './evaluate/code';
import { TypedTemplate } from './type-resolution';
import { loadTypeSystem, readTemplate, stackNameFromFileName } from './util';

let verbosity = 0;

async function main() {
  const argv = await yargs
    .usage(
      '$0 <filename>',
      'Disassemble a CDK app from a declarative JSON or YAML template'
    )
    .option('out', {
      alias: 'o',
      type: 'string',
      demandOption: true,
      describe: 'The directory path to output the disassemble CDK app.',
    })
    .option('verbose', {
      alias: 'v',
      type: 'count',
      description: 'Show debug output. Repeat to increase verbosity.',
    })
    .positional('filename', { type: 'string', required: true }).argv;

  verbosity = argv.verbose;

  const templateFile = argv.filename;
  if (!templateFile) {
    throw new Error('filename is missing');
  }
  const outPath = argv.out;

  const parsedTemplate = await readTemplate(templateFile);
  const stackName = stackNameFromFileName(templateFile);
  const typeSystem = await loadTypeSystem();

  const annotations = AnnotationsContext.root();
  const template = new TypedTemplate(parsedTemplate, {
    annotations,
    typeSystem,
  });

  const context = new CodeEvaluationContext({
    stack: new cdk.Stack(undefined, stackName),
    template,
    typeSystem,
    outPath,
    stackName,
  });
  const ev = new CodeEvaluator(context);

  const project = ev.evaluateTemplate(annotations);

  if (annotations.hasErrors()) {
    throw new DeclarativeStackError(annotations);
  }

  project.saveSync();
}

main().catch((e) => {
  if (e instanceof DeclarativeStackError) {
    e = e.toString(verbosity >= 1);
  }
  // eslint-disable-next-line no-console
  console.error(chalk.red(e));
  process.exit(1);
});
