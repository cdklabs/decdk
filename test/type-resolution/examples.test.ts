import * as reflect from 'jsii-reflect';
import { readTemplate } from '../../src';
import { resolveResourceLike } from '../../src/type-resolution';
import { testExamples, Testing } from '../util';

let typeSystem: reflect.TypeSystem;
beforeAll(async () => {
  typeSystem = await Testing.typeSystem;
});

testExamples(async (example) => {
  const template = await readTemplate(example.path);

  const typedTemplate = template
    .resourceGraph()
    .map((_, resource) => resolveResourceLike(resource, typeSystem));

  expect(template.template).toBeValidTemplate();
  expect(typedTemplate).toMatchSnapshot();
});
