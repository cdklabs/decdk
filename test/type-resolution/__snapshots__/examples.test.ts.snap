// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apigw.json 1`] = `
DependencyGraph {
  "_dependencies": Map {
    "HelloLambda" => Set {},
    "MyApi" => Set {
      "HelloLambda",
    },
    "GetRoot" => Set {
      "MyApi",
    },
  },
  "keys": Set {
    "HelloLambda",
    "MyApi",
    "GetRoot",
  },
  "nodes": Object {
    "GetRoot": Object {
      "fqn": "aws-cdk-lib.aws_apigateway.Method",
      "logicalId": "GetRoot",
      "namespace": "aws_apigateway",
      "props": Object {
        "fields": Object {
          "httpMethod": Object {
            "type": "string",
            "value": "GET",
          },
          "resource": Object {
            "attribute": Object {
              "type": "string",
              "value": "root",
            },
            "fn": "getAtt",
            "logicalId": "MyApi",
            "type": "intrinsic",
          },
        },
        "type": "struct",
      },
      "tags": Array [],
      "type": "construct",
    },
    "HelloLambda": Object {
      "fqn": "aws-cdk-lib.aws_lambda.Function",
      "logicalId": "HelloLambda",
      "namespace": "aws_lambda",
      "props": Object {
        "fields": Object {
          "code": Object {
            "args": Object {
              "array": Array [
                Object {
                  "type": "string",
                  "value": "examples/lambda-handler",
                },
                Object {
                  "fields": Object {},
                  "type": "struct",
                },
              ],
              "type": "array",
            },
            "fqn": "aws-cdk-lib.aws_lambda.Code",
            "method": "fromAsset",
            "namespace": "aws_lambda",
            "type": "staticMethodCall",
          },
          "handler": Object {
            "type": "string",
            "value": "index.handler",
          },
          "runtime": Object {
            "fqn": "aws-cdk-lib.aws_lambda.Runtime",
            "namespace": "aws_lambda",
            "property": "PYTHON_3_6",
            "type": "staticProperty",
          },
        },
        "type": "struct",
      },
      "tags": Array [
        Object {
          "key": "category",
          "value": "ecommerce",
        },
      ],
      "type": "construct",
    },
    "MyApi": Object {
      "fqn": "aws-cdk-lib.aws_apigateway.LambdaRestApi",
      "logicalId": "MyApi",
      "namespace": "aws_apigateway",
      "props": Object {
        "fields": Object {
          "handler": Object {
            "fn": "ref",
            "logicalId": "HelloLambda",
            "type": "intrinsic",
          },
        },
        "type": "struct",
      },
      "tags": Array [],
      "type": "construct",
    },
  },
}
`;

exports[`ecs.json 1`] = `
DependencyGraph {
  "_dependencies": Map {
    "VPC" => Set {},
    "Cluster" => Set {
      "VPC",
    },
    "MyTaskDef" => Set {},
    "ContainerDef" => Set {
      "MyTaskDef",
    },
    "Service" => Set {
      "Cluster",
      "MyTaskDef",
    },
  },
  "keys": Set {
    "VPC",
    "MyTaskDef",
    "Cluster",
    "ContainerDef",
    "Service",
  },
  "nodes": Object {
    "Cluster": Object {
      "fqn": "aws-cdk-lib.aws_ecs.Cluster",
      "logicalId": "Cluster",
      "namespace": "aws_ecs",
      "props": Object {
        "fields": Object {
          "vpc": Object {
            "fn": "ref",
            "logicalId": "VPC",
            "type": "intrinsic",
          },
        },
        "type": "struct",
      },
      "tags": Array [],
      "type": "construct",
    },
    "ContainerDef": Object {
      "fqn": "aws-cdk-lib.aws_ecs.ContainerDefinition",
      "logicalId": "ContainerDef",
      "namespace": "aws_ecs",
      "props": Object {
        "fields": Object {
          "essential": Object {
            "type": "boolean",
            "value": true,
          },
          "image": Object {
            "args": Object {
              "array": Array [
                Object {
                  "type": "string",
                  "value": "redis",
                },
                Object {
                  "fields": Object {},
                  "type": "struct",
                },
              ],
              "type": "array",
            },
            "fqn": "aws-cdk-lib.aws_ecs.ContainerImage",
            "method": "fromRegistry",
            "namespace": "aws_ecs",
            "type": "staticMethodCall",
          },
          "memoryLimitMiB": Object {
            "type": "number",
            "value": 1024,
          },
          "taskDefinition": Object {
            "fn": "ref",
            "logicalId": "MyTaskDef",
            "type": "intrinsic",
          },
        },
        "type": "struct",
      },
      "tags": Array [],
      "type": "construct",
    },
    "MyTaskDef": Object {
      "fqn": "aws-cdk-lib.aws_ecs.TaskDefinition",
      "logicalId": "MyTaskDef",
      "namespace": "aws_ecs",
      "props": Object {
        "fields": Object {
          "compatibility": Object {
            "choice": "FARGATE",
            "fqn": "aws-cdk-lib.aws_ecs.Compatibility",
            "namespace": "aws_ecs",
            "type": "enum",
          },
          "cpu": Object {
            "type": "string",
            "value": "1024",
          },
          "family": Object {
            "type": "string",
            "value": "redis",
          },
          "memoryMiB": Object {
            "type": "string",
            "value": "1GB",
          },
          "networkMode": Object {
            "choice": "AWS_VPC",
            "fqn": "aws-cdk-lib.aws_ecs.NetworkMode",
            "namespace": "aws_ecs",
            "type": "enum",
          },
        },
        "type": "struct",
      },
      "tags": Array [],
      "type": "construct",
    },
    "Service": Object {
      "fqn": "aws-cdk-lib.aws_ecs.FargateService",
      "logicalId": "Service",
      "namespace": "aws_ecs",
      "props": Object {
        "fields": Object {
          "cluster": Object {
            "fn": "ref",
            "logicalId": "Cluster",
            "type": "intrinsic",
          },
          "taskDefinition": Object {
            "fn": "ref",
            "logicalId": "MyTaskDef",
            "type": "intrinsic",
          },
        },
        "type": "struct",
      },
      "tags": Array [],
      "type": "construct",
    },
    "VPC": Object {
      "fqn": "aws-cdk-lib.aws_ec2.Vpc",
      "logicalId": "VPC",
      "namespace": "aws_ec2",
      "props": Object {
        "fields": Object {
          "maxAzs": Object {
            "type": "number",
            "value": 1,
          },
        },
        "type": "struct",
      },
      "tags": Array [],
      "type": "construct",
    },
  },
}
`;

exports[`lambda-events.json 1`] = `
DependencyGraph {
  "_dependencies": Map {
    "MyTopic" => Set {},
    "Table" => Set {},
    "HelloWorldFunction" => Set {
      "Table",
      "MyTopic",
    },
  },
  "keys": Set {
    "MyTopic",
    "Table",
    "HelloWorldFunction",
  },
  "nodes": Object {
    "HelloWorldFunction": Object {
      "fqn": "aws-cdk-lib.aws_lambda.Function",
      "logicalId": "HelloWorldFunction",
      "namespace": "aws_lambda",
      "props": Object {
        "fields": Object {
          "code": Object {
            "args": Object {
              "array": Array [
                Object {
                  "type": "string",
                  "value": "examples/lambda-handler",
                },
                Object {
                  "fields": Object {},
                  "type": "struct",
                },
              ],
              "type": "array",
            },
            "fqn": "aws-cdk-lib.aws_lambda.Code",
            "method": "fromAsset",
            "namespace": "aws_lambda",
            "type": "staticMethodCall",
          },
          "environment": Object {
            "fields": Object {
              "Param": Object {
                "type": "string",
                "value": "f",
              },
            },
            "type": "object",
          },
          "events": Object {
            "array": Array [
              Object {
                "args": Object {
                  "array": Array [
                    Object {
                      "fn": "ref",
                      "logicalId": "Table",
                      "type": "intrinsic",
                    },
                    Object {
                      "fields": Object {
                        "startingPosition": Object {
                          "choice": "TRIM_HORIZON",
                          "fqn": "aws-cdk-lib.aws_lambda.StartingPosition",
                          "namespace": "aws_lambda",
                          "type": "enum",
                        },
                      },
                      "type": "struct",
                    },
                  ],
                  "type": "array",
                },
                "fqn": "aws-cdk-lib.aws_lambda_event_sources.DynamoEventSource",
                "namespace": "aws_lambda_event_sources",
                "type": "initializer",
              },
              Object {
                "args": Object {
                  "array": Array [
                    Object {
                      "type": "string",
                      "value": "GET",
                    },
                    Object {
                      "type": "string",
                      "value": "/hello",
                    },
                    Object {
                      "fields": Object {},
                      "type": "struct",
                    },
                  ],
                  "type": "array",
                },
                "fqn": "aws-cdk-lib.aws_lambda_event_sources.ApiEventSource",
                "namespace": "aws_lambda_event_sources",
                "type": "initializer",
              },
              Object {
                "args": Object {
                  "array": Array [
                    Object {
                      "type": "string",
                      "value": "POST",
                    },
                    Object {
                      "type": "string",
                      "value": "/hello",
                    },
                    Object {
                      "fields": Object {},
                      "type": "struct",
                    },
                  ],
                  "type": "array",
                },
                "fqn": "aws-cdk-lib.aws_lambda_event_sources.ApiEventSource",
                "namespace": "aws_lambda_event_sources",
                "type": "initializer",
              },
              Object {
                "args": Object {
                  "array": Array [
                    Object {
                      "fn": "ref",
                      "logicalId": "MyTopic",
                      "type": "intrinsic",
                    },
                    Object {
                      "fields": Object {},
                      "type": "struct",
                    },
                  ],
                  "type": "array",
                },
                "fqn": "aws-cdk-lib.aws_lambda_event_sources.SnsEventSource",
                "namespace": "aws_lambda_event_sources",
                "type": "initializer",
              },
            ],
            "type": "array",
          },
          "handler": Object {
            "type": "string",
            "value": "app.hello_handler",
          },
          "runtime": Object {
            "fqn": "aws-cdk-lib.aws_lambda.Runtime",
            "namespace": "aws_lambda",
            "property": "PYTHON_3_6",
            "type": "staticProperty",
          },
        },
        "type": "struct",
      },
      "tags": Array [],
      "type": "construct",
    },
    "MyTopic": Object {
      "fqn": "aws-cdk-lib.aws_sns.Topic",
      "logicalId": "MyTopic",
      "namespace": "aws_sns",
      "props": Object {
        "fields": Object {},
        "type": "struct",
      },
      "tags": Array [],
      "type": "construct",
    },
    "Table": Object {
      "fqn": "aws-cdk-lib.aws_dynamodb.Table",
      "logicalId": "Table",
      "namespace": "aws_dynamodb",
      "props": Object {
        "fields": Object {
          "partitionKey": Object {
            "fields": Object {
              "name": Object {
                "type": "string",
                "value": "ID",
              },
              "type": Object {
                "choice": "STRING",
                "fqn": "aws-cdk-lib.aws_dynamodb.AttributeType",
                "namespace": "aws_dynamodb",
                "type": "enum",
              },
            },
            "type": "struct",
          },
          "stream": Object {
            "choice": "NEW_AND_OLD_IMAGES",
            "fqn": "aws-cdk-lib.aws_dynamodb.StreamViewType",
            "namespace": "aws_dynamodb",
            "type": "enum",
          },
        },
        "type": "struct",
      },
      "tags": Array [],
      "type": "construct",
    },
  },
}
`;

exports[`lambda-layer.json 1`] = `
DependencyGraph {
  "_dependencies": Map {
    "AwsCliLayer" => Set {},
    "Lambda" => Set {
      "AwsCliLayer",
    },
  },
  "keys": Set {
    "AwsCliLayer",
    "Lambda",
  },
  "nodes": Object {
    "AwsCliLayer": Object {
      "fqn": "aws-cdk-lib.lambda_layer_awscli.AwsCliLayer",
      "logicalId": "AwsCliLayer",
      "namespace": "lambda_layer_awscli",
      "props": Object {
        "type": "void",
      },
      "tags": Array [],
      "type": "construct",
    },
    "Lambda": Object {
      "fqn": "aws-cdk-lib.aws_lambda.Function",
      "logicalId": "Lambda",
      "namespace": "aws_lambda",
      "props": Object {
        "fields": Object {
          "code": Object {
            "args": Object {
              "array": Array [
                Object {
                  "type": "string",
                  "value": "exports.handler = async function() { return 'SUCCESS'; }",
                },
              ],
              "type": "array",
            },
            "fqn": "aws-cdk-lib.aws_lambda.Code",
            "method": "fromInline",
            "namespace": "aws_lambda",
            "type": "staticMethodCall",
          },
          "handler": Object {
            "type": "string",
            "value": "index.handler",
          },
          "layers": Object {
            "array": Array [
              Object {
                "fn": "ref",
                "logicalId": "AwsCliLayer",
                "type": "intrinsic",
              },
            ],
            "type": "array",
          },
          "runtime": Object {
            "fqn": "aws-cdk-lib.aws_lambda.Runtime",
            "namespace": "aws_lambda",
            "property": "NODEJS_16_X",
            "type": "staticProperty",
          },
        },
        "type": "struct",
      },
      "tags": Array [],
      "type": "construct",
    },
  },
}
`;

exports[`lambda-role-removed.json 1`] = `
DependencyGraph {
  "_dependencies": Map {
    "Lambda" => Set {},
  },
  "keys": Set {
    "Lambda",
  },
  "nodes": Object {
    "Lambda": Object {
      "fqn": "aws-cdk-lib.aws_lambda.Function",
      "logicalId": "Lambda",
      "namespace": "aws_lambda",
      "props": Object {
        "fields": Object {
          "code": Object {
            "args": Object {
              "array": Array [
                Object {
                  "type": "string",
                  "value": "examples/lambda-handler",
                },
                Object {
                  "fields": Object {},
                  "type": "struct",
                },
              ],
              "type": "array",
            },
            "fqn": "aws-cdk-lib.aws_lambda.Code",
            "method": "fromAsset",
            "namespace": "aws_lambda",
            "type": "staticMethodCall",
          },
          "description": Object {
            "type": "string",
            "value": "This function has its service role removed, for demonstration purposes. You probably wouldn't do this to a real function",
          },
          "handler": Object {
            "type": "string",
            "value": "index.handler",
          },
          "runtime": Object {
            "fqn": "aws-cdk-lib.aws_lambda.Runtime",
            "namespace": "aws_lambda",
            "property": "NODEJS",
            "type": "staticProperty",
          },
        },
        "type": "struct",
      },
      "tags": Array [],
      "type": "construct",
    },
  },
}
`;

exports[`lambda-topic.json 1`] = `
DependencyGraph {
  "_dependencies": Map {
    "Topic" => Set {},
    "Lambda" => Set {
      "Topic",
    },
  },
  "keys": Set {
    "Topic",
    "Lambda",
  },
  "nodes": Object {
    "Lambda": Object {
      "fqn": "aws-cdk-lib.aws_lambda.Function",
      "logicalId": "Lambda",
      "namespace": "aws_lambda",
      "props": Object {
        "fields": Object {
          "code": Object {
            "args": Object {
              "array": Array [
                Object {
                  "type": "string",
                  "value": "examples/lambda-handler",
                },
                Object {
                  "fields": Object {},
                  "type": "struct",
                },
              ],
              "type": "array",
            },
            "fqn": "aws-cdk-lib.aws_lambda.Code",
            "method": "fromAsset",
            "namespace": "aws_lambda",
            "type": "staticMethodCall",
          },
          "events": Object {
            "array": Array [
              Object {
                "args": Object {
                  "array": Array [
                    Object {
                      "fn": "ref",
                      "logicalId": "Topic",
                      "type": "intrinsic",
                    },
                    Object {
                      "fields": Object {},
                      "type": "struct",
                    },
                  ],
                  "type": "array",
                },
                "fqn": "aws-cdk-lib.aws_lambda_event_sources.SnsEventSource",
                "namespace": "aws_lambda_event_sources",
                "type": "initializer",
              },
            ],
            "type": "array",
          },
          "handler": Object {
            "type": "string",
            "value": "index.handler",
          },
          "runtime": Object {
            "fqn": "aws-cdk-lib.aws_lambda.Runtime",
            "namespace": "aws_lambda",
            "property": "NODEJS",
            "type": "staticProperty",
          },
        },
        "type": "struct",
      },
      "tags": Array [],
      "type": "construct",
    },
    "Topic": Object {
      "fqn": "aws-cdk-lib.aws_sns.Topic",
      "logicalId": "Topic",
      "namespace": "aws_sns",
      "props": Object {
        "fields": Object {},
        "type": "struct",
      },
      "tags": Array [],
      "type": "construct",
    },
  },
}
`;

exports[`pipeline.json 1`] = `
DependencyGraph {
  "_dependencies": Map {
    "Repo" => Set {
      "Key",
      "BuildProject",
    },
    "BuildProject" => Set {
      "Key",
    },
    "Key" => Set {},
    "Pipeline" => Set {
      "Repo",
      "BuildProject",
    },
  },
  "keys": Set {
    "Key",
    "BuildProject",
    "Repo",
    "Pipeline",
  },
  "nodes": Object {
    "BuildProject": Object {
      "fqn": "aws-cdk-lib.aws_codebuild.PipelineProject",
      "logicalId": "BuildProject",
      "namespace": "aws_codebuild",
      "props": Object {
        "fields": Object {
          "encryptionKey": Object {
            "fn": "ref",
            "logicalId": "Key",
            "type": "intrinsic",
          },
        },
        "type": "struct",
      },
      "tags": Array [],
      "type": "construct",
    },
    "Key": Object {
      "fqn": "aws-cdk-lib.aws_kms.Key",
      "logicalId": "Key",
      "namespace": "aws_kms",
      "props": Object {
        "fields": Object {},
        "type": "struct",
      },
      "tags": Array [],
      "type": "construct",
    },
    "Pipeline": Object {
      "fqn": "aws-cdk-lib.aws_codepipeline.Pipeline",
      "logicalId": "Pipeline",
      "namespace": "aws_codepipeline",
      "props": Object {
        "fields": Object {
          "stages": Object {
            "array": Array [
              Object {
                "fields": Object {
                  "actions": Object {
                    "array": Array [
                      Object {
                        "args": Object {
                          "array": Array [
                            Object {
                              "fields": Object {
                                "actionName": Object {
                                  "type": "string",
                                  "value": "Source",
                                },
                                "output": Object {
                                  "args": Object {
                                    "array": Array [
                                      Object {
                                        "type": "string",
                                        "value": "Source",
                                      },
                                    ],
                                    "type": "array",
                                  },
                                  "fqn": "aws-cdk-lib.aws_codepipeline.Artifact",
                                  "method": "artifact",
                                  "namespace": "aws_codepipeline",
                                  "type": "staticMethodCall",
                                },
                                "repository": Object {
                                  "fn": "ref",
                                  "logicalId": "Repo",
                                  "type": "intrinsic",
                                },
                              },
                              "type": "struct",
                            },
                          ],
                          "type": "array",
                        },
                        "fqn": "aws-cdk-lib.aws_codepipeline_actions.CodeCommitSourceAction",
                        "namespace": "aws_codepipeline_actions",
                        "type": "initializer",
                      },
                    ],
                    "type": "array",
                  },
                  "stageName": Object {
                    "type": "string",
                    "value": "Source",
                  },
                },
                "type": "struct",
              },
              Object {
                "fields": Object {
                  "actions": Object {
                    "array": Array [
                      Object {
                        "args": Object {
                          "array": Array [
                            Object {
                              "fields": Object {
                                "actionName": Object {
                                  "type": "string",
                                  "value": "Build",
                                },
                                "input": Object {
                                  "args": Object {
                                    "array": Array [
                                      Object {
                                        "type": "string",
                                        "value": "Source",
                                      },
                                    ],
                                    "type": "array",
                                  },
                                  "fqn": "aws-cdk-lib.aws_codepipeline.Artifact",
                                  "method": "artifact",
                                  "namespace": "aws_codepipeline",
                                  "type": "staticMethodCall",
                                },
                                "outputs": Object {
                                  "array": Array [
                                    Object {
                                      "args": Object {
                                        "array": Array [
                                          Object {
                                            "type": "string",
                                            "value": "Build",
                                          },
                                        ],
                                        "type": "array",
                                      },
                                      "fqn": "aws-cdk-lib.aws_codepipeline.Artifact",
                                      "method": "artifact",
                                      "namespace": "aws_codepipeline",
                                      "type": "staticMethodCall",
                                    },
                                  ],
                                  "type": "array",
                                },
                                "project": Object {
                                  "fn": "ref",
                                  "logicalId": "BuildProject",
                                  "type": "intrinsic",
                                },
                              },
                              "type": "struct",
                            },
                          ],
                          "type": "array",
                        },
                        "fqn": "aws-cdk-lib.aws_codepipeline_actions.CodeBuildAction",
                        "namespace": "aws_codepipeline_actions",
                        "type": "initializer",
                      },
                    ],
                    "type": "array",
                  },
                  "stageName": Object {
                    "type": "string",
                    "value": "Build",
                  },
                },
                "type": "struct",
              },
              Object {
                "fields": Object {
                  "actions": Object {
                    "array": Array [
                      Object {
                        "args": Object {
                          "array": Array [
                            Object {
                              "fields": Object {
                                "actionName": Object {
                                  "type": "string",
                                  "value": "Deploy",
                                },
                                "adminPermissions": Object {
                                  "type": "boolean",
                                  "value": true,
                                },
                                "stackName": Object {
                                  "type": "string",
                                  "value": "MyStack",
                                },
                                "templatePath": Object {
                                  "args": Object {
                                    "array": Array [
                                      Object {
                                        "type": "string",
                                        "value": "Build",
                                      },
                                      Object {
                                        "type": "string",
                                        "value": "template.yaml",
                                      },
                                    ],
                                    "type": "array",
                                  },
                                  "fqn": "aws-cdk-lib.aws_codepipeline.ArtifactPath",
                                  "method": "artifactPath",
                                  "namespace": "aws_codepipeline",
                                  "type": "staticMethodCall",
                                },
                              },
                              "type": "struct",
                            },
                          ],
                          "type": "array",
                        },
                        "fqn": "aws-cdk-lib.aws_codepipeline_actions.CloudFormationCreateUpdateStackAction",
                        "namespace": "aws_codepipeline_actions",
                        "type": "initializer",
                      },
                    ],
                    "type": "array",
                  },
                  "stageName": Object {
                    "type": "string",
                    "value": "Deploy",
                  },
                },
                "type": "struct",
              },
            ],
            "type": "array",
          },
        },
        "type": "struct",
      },
      "tags": Array [],
      "type": "construct",
    },
    "Repo": Object {
      "fqn": "aws-cdk-lib.aws_codecommit.Repository",
      "logicalId": "Repo",
      "namespace": "aws_codecommit",
      "props": Object {
        "fields": Object {
          "repositoryName": Object {
            "type": "string",
            "value": "my-first-decdk-repo",
          },
        },
        "type": "struct",
      },
      "tags": Array [],
      "type": "construct",
    },
  },
}
`;

exports[`pure-cfn.json 1`] = `
DependencyGraph {
  "_dependencies": Map {
    "Hello" => Set {},
    "MyLogGroup" => Set {},
  },
  "keys": Set {
    "Hello",
    "MyLogGroup",
  },
  "nodes": Object {
    "Hello": Object {
      "fqn": "aws-cdk-lib.aws_sqs.Queue",
      "logicalId": "Hello",
      "namespace": "aws_sqs",
      "props": Object {
        "fields": Object {
          "encryption": Object {
            "choice": "UNENCRYPTED",
            "fqn": "aws-cdk-lib.aws_sqs.QueueEncryption",
            "namespace": "aws_sqs",
            "type": "enum",
          },
        },
        "type": "struct",
      },
      "tags": Array [],
      "type": "construct",
    },
    "MyLogGroup": Object {
      "fqn": "aws-cdk-lib.CfnResource",
      "logicalId": "MyLogGroup",
      "namespace": undefined,
      "props": Object {
        "fields": Object {
          "properties": Object {
            "fields": Object {
              "LogGroupName": Object {
                "type": "any",
                "value": Object {
                  "fn": "ref",
                  "logicalId": "AWS::AccountId",
                  "type": "intrinsic",
                },
              },
            },
            "type": "object",
          },
          "type": Object {
            "type": "string",
            "value": "AWS::Logs::LogGroup",
          },
        },
        "type": "struct",
      },
      "tags": Array [],
      "type": "resource",
    },
  },
}
`;

exports[`queue-kms.json 1`] = `
DependencyGraph {
  "_dependencies": Map {
    "MyQueue" => Set {},
  },
  "keys": Set {
    "MyQueue",
  },
  "nodes": Object {
    "MyQueue": Object {
      "fqn": "aws-cdk-lib.aws_sqs.Queue",
      "logicalId": "MyQueue",
      "namespace": "aws_sqs",
      "props": Object {
        "fields": Object {
          "encryption": Object {
            "choice": "KMS",
            "fqn": "aws-cdk-lib.aws_sqs.QueueEncryption",
            "namespace": "aws_sqs",
            "type": "enum",
          },
        },
        "type": "struct",
      },
      "tags": Array [],
      "type": "construct",
    },
  },
}
`;

exports[`vpc.json 1`] = `
DependencyGraph {
  "_dependencies": Map {
    "VPC" => Set {},
  },
  "keys": Set {
    "VPC",
  },
  "nodes": Object {
    "VPC": Object {
      "fqn": "aws-cdk-lib.aws_ec2.Vpc",
      "logicalId": "VPC",
      "namespace": "aws_ec2",
      "props": Object {
        "fields": Object {},
        "type": "struct",
      },
      "tags": Array [],
      "type": "construct",
    },
  },
}
`;
